contrib/win32/win32compat/signal_wait.c

```diff
@@ -0,0 +1,194 @@
+	waitstruct->return_value = WaitForMultipleObjectsEx(waitstruct->num_handles,
+		waitstruct->handles, FALSE, INFINITE, TRUE);
+
+	/* notify the main thread that an event was found */
+	SetEvent(waitstruct->wait_event);
+
+	return TRUE;
+}
+
+VOID CALLBACK wait_for_multiple_objects_interrupter(_In_ ULONG_PTR dwParam)
+{
+	/* we must explicitly exit the thread since the thread could have been received 
+	 * the alert prior to the thread running in which case its acknowledged when  
+	 * the threads starts running instead of when its waiting at 
+	 * WaitForMultipleObjectsEx */
+	ExitThread(0);
+}
+
+DWORD wait_for_multiple_objects_enhanced(_In_ DWORD  nCount, _In_ const HANDLE *lpHandles,
+	_In_ DWORD dwMilliseconds, _In_ BOOL bAlertable)
+{
+	/* if less than the normal maximum then just use the built-in function
+	 * to avoid the overhead of another thread */
+	if (nCount <= MAXIMUM_WAIT_OBJECTS) {
+
+		DWORD wait_ret = WaitForMultipleObjectsEx(nCount, lpHandles,
+			FALSE, dwMilliseconds, bAlertable);
+
+		if (wait_ret == WAIT_IO_COMPLETION) return WAIT_IO_COMPLETION_ENHANCED;
+		if (wait_ret == WAIT_TIMEOUT) return WAIT_TIMEOUT_ENHANCED;
+
+		/* translate normal offset to enhanced offset for abandoned threads */
+		if (wait_ret >= WAIT_ABANDONED_0 && wait_ret < WAIT_ABANDONED_0 + MAXIMUM_WAIT_OBJECTS) {
+			return WAIT_ABANDONED_0_ENHANCED + (wait_ret - WAIT_ABANDONED_0);
+		}
+
+		/* translate normal offset to enhanced offset for signaled threads */
+		if (wait_ret >= WAIT_OBJECT_0 && wait_ret < WAIT_OBJECT_0 + MAXIMUM_WAIT_OBJECTS) {
+			return WAIT_OBJECT_0_ENHANCED + (wait_ret - WAIT_OBJECT_0);
+		}
+
+		return WAIT_FAILED_ENHANCED;
+	}
+
+	/* number of separate bins / threads required to create */
+	const DWORD bin_size = MAXIMUM_WAIT_OBJECTS; 
+	const DWORD bins_total = (nCount - 1) / bin_size + 1;
+
+	/* setup synchronization variables */
+	HANDLE wait_event = CreateEvent(NULL, TRUE, FALSE, NULL);
+
+	/* allocate an area to communicate with our threads */
+	wait_for_multiple_objects_struct * wait_bins = (wait_for_multiple_objects_struct *)
+		calloc(bins_total, sizeof(wait_for_multiple_objects_struct));
+
+	/* initialize each thread handling bin */
+	for (DWORD bin = 0; bin < bins_total; bin++) {
+
+		const int handles_processed = bin * bin_size;
+
+		wait_bins[bin].return_value = WAIT_FAILED - 1;
+		wait_bins[bin].wait_event = wait_event;
+		wait_bins[bin].handles = &(lpHandles[handles_processed]);
+		wait_bins[bin].num_handles = min(nCount - handles_processed, bin_size);
+
+		/* create a thread for this bin */
+		if ((wait_bins[bin].thread_handle = CreateThread(NULL, 2048,
+			wait_for_multiple_objects_thread, (LPVOID) &(wait_bins[bin]), 0, NULL)) == NULL) {
+			goto cleanup;
+		}
+	}
+
+	/* wait for at least one thread to return */
+	DWORD return_value = WAIT_FAILED_ENHANCED;
+	DWORD wait_ret = WaitForSingleObjectEx(wait_event, dwMilliseconds, bAlertable);
+
```

Comment:
You should issue APCs on worker threads right here and wait for all of them to terminate. This needs to be done before you read "bins" state below (to synchronize writes and reads on this state).

Reply:
The only potential conflict value would be return_value which would be an aligned 32-bit value.  Those read/writes will always be atomic and at least one of them will be set before SetEvent() is called.   Am I overlooking another potential concern?

Reply:
No. There isn't any potential concern yet. From a logical stand point, once the main thread woke up, there is no point keeping the worker threads active. If we abort and clean them up right away, the rest of the trailing code need not make any assumptions of what else might be running in parallel. This will help us keep the logic simple, easy to maintain and debug (in case of any issues).